distribution/lib/Standard/Base/0.0.0-dev/src/Data/XML.enso

```diff
@@ -0,0 +1,385 @@
+
+type XML_Document
+    ## Read an XML document from a file.
+
+       Arguments:
+       - file: the `File` to read the XML document from.
+
+       If there is an error reading the file, `File_Error` is thrown.
+
+       If there is a parsing error, `XML_Error.Parse_Error` is thrown.
+
+       > Example
+         Read an XML document in 'doc.xml'.
+
+             file = enso_project.data / "doc.xml"
+             doc = XML_Document.from_file test_file
+    from_file : File -> XML_Document ! XML_Error | File_Error
+    from_file file:File =
+        File_Error.handle_java_exceptions file <| XML_Error.handle_java_exceptions <|
+            file.with_input_stream [File_Access.Read] XML_Document.from_stream
+
+    ## Read an XML document from an input stream.
+
+       Arguments:
+       - input_stream: the input stread to read the XML document from.
+
+       If there is a parsing error, `XML_Error.Parse_Error` is thrown.
+
+       > Example
+         Read an XML document from an input_stream.
+
+             doc = XML_Document.from_stream input_stream
+    from_stream : Input_Stream -> XML_Document ! XML_Error
+    from_stream input_stream:Input_Stream =
+        XML_Error.handle_java_exceptions <|
+           input_stream.with_java_stream java_stream-> XML_Document.from_source java_stream
+
+    ## Read an XML document from a string.
+
+       Arguments:
+       - xml_string: The string to read the XML document from.
+
+       If there is a parsing error, `XML_Error.Parse_Error` is thrown.
+
+       > Example
+         Read an XML document from an string.
+
+             xml_string = "<foo></foo>"
+             doc = XML_Document.from_string xml_string
+    from_string : Text -> XML_Document ! XML_Error
+    from_string xml_string:Text =
+        XML_Error.handle_java_exceptions <|
+            string_reader = StringReader.new xml_string
+            XML_Document.from_source (InputSource.new string_reader)
+
+    ## PRIVATE
+       Read XML from an input source.
+    from_source : Any -> XML_Document ! XML_Error
+    from_source input_source =
+        document_builder_factory = DocumentBuilderFactory.newInstance
+        document_builder = document_builder_factory.newDocumentBuilder
+        XML_Document.Value (document_builder.parse input_source)
+
+    ## Get the root element of the document.
+
+       > Example
+         Get the root element of a document.
+
+             doc = XML_Document.from_file file
+             root = doc.root_element
+    root_element : XML_Element ! XML_Error
+    root_element self =
+        XML_Error.handle_java_exceptions <|
+            java_element = self.java_document.getDocumentElement
+            XML_Element.Value java_element
+
+    ## PRIVATE
+    Value (java_document:Document)
+
+type XML_Element
+    ## Gets the tag of an xml element.
+
+       Example:
+           XML_Document.from_string '<foo>hello</foo>' . root_element . at 0 . tag
+           # => "foo"
```

Comment:
I think I don't understand, why is the `at 0` needed here??

Also the method is `name`, but the example says `tag`. This is inconsistent.

Shouldn't that be
```suggestion
           XML_Document.from_string '<foo>hello</foo>' . root_element . name
           # => "foo"
```

Reply:
Ideally, it would be best to add tests for each of the examples to verify that they behave 'as advertised'.

@jdunkerley the more examples we have, the harder it will be to keep them in sync. There used to be efforts to be able to _extract and run_ examples from the docstrings, to be able to validate them - that they compile but we also could validate if they return what is expected (e.g. by comparing the `to_text` of the result). Do you think we should revive these efforst soon(ish)?

Reply:
I would personally love to revive this. I'll add a task.

Reply:
https://github.com/enso-org/enso/issues/7949